bitswap/provider.go

```diff
@@ -94,264 +121,312 @@ func NewProvider(
+		if err := provider.network.SendMessage(ctx, peerID, msg); err != nil {
+			log.Warnf("Failed to send message to %s: %v", peerID, err)
+			provider.responseQueue.TasksDone(peerID, tasks...)
+		}
 
-func (provider *Provider) signalWork() {
-	// Request work if any worker isn't running or do nothing otherwise
-	select {
-	case provider.workReady <- struct{}{}:
-	default:
+		provider.responseQueue.TasksDone(peerID, tasks...)
+		log.Debugf("Sent message to %s", peerID)
 	}
 }
 
-func (provider *Provider) runWorker() {
-	for {
-		peer, tasks, pending := provider.taskQueue.PopTasks(targetMessageSize)
+func (provider *Provider) handleRetrievals() {
+	ctx := context.Background()
 
-		// If there's nothing to do, wait for something to happen
+	for {
+		peerID, tasks, _ := provider.retrievalQueue.PopTasks(1)
 		if len(tasks) == 0 {
-			select {
-			case _, ok := <-provider.workReady:
-				// If the workReady channel closed, stop the worker
-				if !ok {
-					logger.Infof("Shutting down worker")
-					return
-				}
-			case <-time.After(250 * time.Millisecond):
-			}
+			time.Sleep(time.Millisecond * 250)
 			continue
 		}
 
-		msg := message.New(false)
+		log.Debugf("Retrieval of %d CIDs queued for %s", len(tasks), peerID)
 
 		for _, task := range tasks {
-			switch topic := task.Topic.(type) {
-			case topicHave:
-				have, err := provider.blockManager.Has(context.Background(), cid.Cid(topic))
-				if err != nil {
-					logger.Errorf("Block load error: %s", err.Error())
-					msg.AddDontHave(cid.Cid(topic))
-				} else if !have {
-					logger.Debugf("Had a block but lost it for want_have: %s", cid.Cid(topic))
-					msg.AddDontHave(cid.Cid(topic))
+			cid, ok := task.Topic.(cid.Cid)
+			if !ok {
+				log.Warnf("Retrieval topic wasn't a CID")
+				continue
+			}
+
+			log.Debugf("Requesting retrieval for %s", cid)
+
+			// Try to start a new retrieval (if it's already running then no
+			// need to error, just continue on to await block)
+			if err := provider.retriever.Request(cid); err != nil {
+				if !errors.As(err, &lassieretriever.ErrRetrievalAlreadyRunning{}) {
+					if errors.Is(err, lassieretriever.ErrNoCandidates) {
+						// Just do a debug print if there were no candidates because this happens a lot
+						log.Debugf("No candidates for %s", cid)
+					} else {
+						// Otherwise, there was a real failure, print with more importance
+						log.Errorf("Request for %s failed: %v", cid, err)
+					}
 				} else {
-					logger.Debugf("Sending want_have to peer: %s", cid.Cid(topic))
-					msg.AddHave(cid.Cid(topic))
+					log.Debugf("Retrieval already running for %s, no new one will be started", cid)
 				}
-			case topicDontHave:
-				msg.AddDontHave(cid.Cid(topic))
-			case topicBlock:
-				blk, err := provider.blockManager.Get(context.Background(), cid.Cid(topic))
+			} else {
+				log.Infof("Started retrieval for %s", cid)
+			}
+
+			// TODO: if retriever.Request() is changed to be blocking, make
```

Comment:
For the record, we're using lassie v0.1.2 which still punts to a goroutine: https://github.com/filecoin-project/lassie/blob/v0.1.2/pkg/retriever/retriever.go#L180

The next update of lassie here will go sync on this call: https://github.com/filecoin-project/lassie/blob/9c4638df01add7d676e53092d1a8d724ba402e59/pkg/retriever/retriever.go#L173

I like the idea of making AwaitBlock cancellable; we'll presumably get the block well before the retrieval has finished (particularly if it's a deep DAG).